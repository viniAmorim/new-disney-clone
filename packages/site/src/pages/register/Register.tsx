import React, { BaseHTMLAttributes } from 'react';
import { UseFormReturn } from 'react-hook-form';
import { Col, Container, Row } from 'react-grid-system';
import { FaPlus } from 'react-icons/fa';

import { FormControl, Input, Heading, Button, Paragraph } from '@site/uikit';
import Link from 'next/link';

import { RegisterFormValues } from './RegisterContainer';
import * as Styled from './styles';

interface RegisterProps extends BaseHTMLAttributes<HTMLDivElement> {
  onRegister: (values: RegisterFormValues) => void;
  form: UseFormReturn;
  isLoading?: boolean;
}

function Register({ onRegister, form, isLoading, ...props }: RegisterProps) {
  return (
    <Styled.Background {...props}>
      <Container>
        <Row>
          <Col sm={6} md={6}>
            <Styled.Logo src="/images/logo.png" alt="Correta.app" />
            <Styled.StyledHeading size="headline">
              Cadastre-se gratuitamente
            </Styled.StyledHeading>
            <form onSubmit={form.handleSubmit(onRegister)}>
              <FormControl
                label="Nome:"
                error={form.formState.errors?.name?.message}
              >
                <Input
                  name="name"
                  placeholder="Seu Nome"
                  {...form.register('name', {
                    required: true,
                    minLength: {
                      value: 5,
                      message: 'Nome é obrigatório',
                    },
                  })}
                />
              </FormControl>
              <FormControl
                label="E-mail:"
                error={form.formState.errors?.email?.message}
              >
                <Input
                  name="email"
                  placeholder="seu@email.com"
                  {...form.register('email', {
                    required: true,
                    pattern: {
                      value: /\S+@\S+\.\S+/,
                      message: 'E-mail inválido.',
                    },
                    minLength: {
                      value: 5,
                      message: 'E-mail é obrigatório',
                    },
                  })}
                />
              </FormControl>
              <FormControl
                label="Senha:"
                error={form.formState.errors?.password?.message}
              >
                <Input
                  name="password"
                  type="password"
                  placeholder="password"
                  {...form.register('password', {
                    required: true,
                    minLength: {
                      value: 5,
                      message: 'Senha é obrigatório',
                    },
                  })}
                  data-testid="password-input"
                />
              </FormControl>
              <FormControl
                label="Repetir Senha:"
                error={form.formState.errors?.repeatPassword?.message}
              >
                <Input
                  name="repeatPassword"
                  type="password"
                  placeholder="password"
                  {...form.register('repeatPassword', {
                    required: true,
                    minLength: {
                      value: 5,
                      message: 'Repetir Senha é obrigatório',
                    },
                    validate: value => {
                      if (value === form.getValues()['password']) {
                        return true;
                      } else {
                        return 'Senha e Repetir senha precisam ser iguais.';
                      }
                    },
                  })}
                  data-testid="repeat-password-input"
                />
              </FormControl>
              <Styled.RegisterContainer>
                <Button
                  data-testid="register-button"
                  type="submit"
                  isLoading={isLoading}
                >
                  Cadastrar
                </Button>
              </Styled.RegisterContainer>
              <Styled.LoginContainer>
                <Paragraph size="small">Já tem cadastro?</Paragraph>
                <Link href="/">
                  <Button variant="ghost">Fazer Login</Button>
                </Link>
              </Styled.LoginContainer>
            </form>
          </Col>
          <Styled.ContentCol sm={6} md={6}>
            <Styled.ContentContainer>
              <img src="/images/home.png" alt="Correta.app" />
              <span>
                <FaPlus />
                <Heading size="display">facilidade</Heading>
              </span>
              <span>
                <FaPlus />
                <Heading size="display">resultados</Heading>
              </span>
              <Paragraph size="headline">
                Tenha acesso e controle de seus cliente, melhorando seu processo
                de vendas e gerando mais resultados.
              </Paragraph>
            </Styled.ContentContainer>
          </Styled.ContentCol>
        </Row>
      </Container>
    </Styled.Background>
  );
}

export default Register;
