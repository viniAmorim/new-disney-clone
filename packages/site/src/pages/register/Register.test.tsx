import React from 'react';

import userEvent from '@testing-library/user-event';

import { fireEvent, render, waitFor } from '~/test/utils';

import Register from './Register';

const onRegister = jest.fn();

describe('Register', () => {
  it('should render correctly', async () => {
    const { queryByTestId } = render(
      <Register data-testid="home" onRegister={onRegister} />,
    );

    const home = queryByTestId('home');

    expect(home).toBeInTheDocument();
    expect(home).toMatchSnapshot();
  });

  it('should show error when name is empty', async () => {
    const { getByTestId, getByText } = render(
      <Register onRegister={onRegister} />,
    );

    const button = getByTestId('register-button');
    fireEvent.click(button);

    await waitFor(() => expect(getByText('Nome é obrigatório')).toBeTruthy());
  });

  it('should show error when email is empty', async () => {
    const { getByTestId, getAllByRole, getByText } = render(
      <Register onRegister={onRegister} />,
    );

    const [nameInput] = getAllByRole('textbox');
    userEvent.type(nameInput, 'Name');

    const button = getByTestId('register-button');
    fireEvent.click(button);

    await waitFor(() => expect(getByText('E-mail é obrigatório')).toBeTruthy());
  });

  it('should show error when email is invalid', async () => {
    const { getAllByRole, getByTestId, getByText } = render(
      <Register onRegister={onRegister} />,
    );

    const [nameInput, emailInput] = getAllByRole('textbox');
    userEvent.type(nameInput, 'Name');
    userEvent.type(emailInput, 'invalid-email.com');

    const button = getByTestId('register-button');
    fireEvent.click(button);

    await waitFor(() => expect(getByText('E-mail inválido')).toBeTruthy());
  });

  it('should show error when password is empty', async () => {
    const { getAllByRole, getByTestId, getByText } = render(
      <Register onRegister={onRegister} />,
    );

    const [nameInput, emailInput] = getAllByRole('textbox');
    userEvent.type(nameInput, 'Name');
    userEvent.type(emailInput, 'email@email.com');

    const button = getByTestId('register-button');
    fireEvent.click(button);

    await waitFor(() => expect(getByText('Senha é obrigatória')).toBeTruthy());
  });

  it('should show error when repeat password is empty', async () => {
    const { getAllByRole, getByTestId, getByText } = render(
      <Register onRegister={onRegister} />,
    );

    const [nameInput, emailInput] = getAllByRole('textbox');
    userEvent.type(nameInput, 'Name');
    userEvent.type(emailInput, 'email@email.com');

    const passwordInput = getByTestId('password-input');
    userEvent.type(passwordInput, 'password');

    const button = getByTestId('register-button');
    fireEvent.click(button);

    await waitFor(() =>
      expect(getByText('Repetir Senha é obrigatória')).toBeTruthy(),
    );
  });

  it('should show error when repeat password is different from password field', async () => {
    const { getAllByRole, getByTestId, getByText } = render(
      <Register onRegister={onRegister} />,
    );

    const [nameInput, emailInput] = getAllByRole('textbox');
    userEvent.type(nameInput, 'Name');
    userEvent.type(emailInput, 'email@email.com');

    const passwordInput = getByTestId('password-input');
    userEvent.type(passwordInput, 'password');

    const repeatPasswordInput = getByTestId('repeat-password-input');
    userEvent.type(repeatPasswordInput, 'password2');

    const button = getByTestId('register-button');
    fireEvent.click(button);

    await waitFor(() =>
      expect(
        getByText('Senha e Repetir senha precisam ser iguais.'),
      ).toBeTruthy(),
    );
  });

  it('should get register data when clicked in register button', async () => {
    const { getAllByRole, getByTestId } = render(
      <Register onRegister={onRegister} />,
    );

    const [nameInput, emailInput] = getAllByRole('textbox');
    userEvent.type(nameInput, 'Name');
    userEvent.type(emailInput, 'email@email.com');

    const passwordInput = getByTestId('password-input');
    userEvent.type(passwordInput, 'password');

    const repeatPasswordInput = getByTestId('repeat-password-input');
    userEvent.type(repeatPasswordInput, 'password');

    const button = getByTestId('register-button');
    fireEvent.click(button);

    await waitFor(() => {
      expect(onRegister.mock.calls[0][0]).toEqual({
        email: 'email@email.com',
        name: 'Name',
        password: 'password',
        repeatPassword: 'password',
      });
    });
  });
});
