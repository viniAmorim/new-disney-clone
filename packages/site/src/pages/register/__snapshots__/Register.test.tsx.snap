// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register should render correctly 1`] = `
<div
  class="styles__Background-sc-f3t2w6-2 A-dCpB"
  data-testid="home"
>
  <div
    style="box-sizing: border-box; position: relative; margin-left: auto; margin-right: auto; padding-left: 15px; padding-right: 15px; max-width: 960px;"
  >
    <div
      style="margin-left: -15px; margin-right: -15px; display: flex; flex-wrap: wrap; flex-grow: 0; flex-shrink: 0; align-items: normal; justify-content: flex-start; flex-direction: row;"
    >
      <div
        style="box-sizing: border-box; min-height: 1px; position: relative; padding-left: 15px; padding-right: 15px; width: 50%; flex-basis: 50%; flex-grow: 0; flex-shrink: 0; max-width: 50%; margin-left: 0%;"
      >
        <img
          alt="Correta.app"
          class="styles__Logo-sc-f3t2w6-0 knMKKh"
          src="/images/logo.png"
        />
        <h1
          class="sc-gKAaRy kbUCYm styles__StyledHeading-sc-f3t2w6-1 hlOYuk"
        >
          Cadastre-se gratuitamente
        </h1>
        <form>
          <div
            class="sc-hKFxyN zFDVE"
          >
            <label
              class="sc-eCApnc CDjog"
            >
              Nome:
            </label>
            <input
              class="sc-iCoGMd kmeLFP"
              name="name"
              placeholder="Seu Nome"
              value=""
            />
          </div>
          <div
            class="sc-hKFxyN zFDVE"
          >
            <label
              class="sc-eCApnc CDjog"
            >
              E-mail:
            </label>
            <input
              class="sc-iCoGMd kmeLFP"
              name="email"
              placeholder="seu@email.com"
              value=""
            />
          </div>
          <div
            class="sc-hKFxyN zFDVE"
          >
            <label
              class="sc-eCApnc CDjog"
            >
              Senha:
            </label>
            <input
              class="sc-iCoGMd kmeLFP"
              data-testid="password-input"
              name="password"
              placeholder="password"
              type="password"
              value=""
            />
          </div>
          <div
            class="sc-hKFxyN zFDVE"
          >
            <label
              class="sc-eCApnc CDjog"
            >
              Repetir Senha:
            </label>
            <input
              class="sc-iCoGMd kmeLFP"
              data-testid="repeat-password-input"
              name="repeatPassword"
              placeholder="password"
              type="password"
              value=""
            />
          </div>
          <div
            class="styles__RegisterContainer-sc-f3t2w6-5 cRgEQs"
          >
            <button
              class="sc-gtsrHT kGzhap"
              data-testid="register-button"
              type="submit"
            >
              Cadastrar
            </button>
          </div>
          <div
            class="styles__LoginContainer-sc-f3t2w6-6 kNfMlu"
          >
            <p
              class="sc-iemWCZ eTebHM"
            >
              JÃ¡ tem cadastro?
            </p>
            <button
              class="sc-gtsrHT gBmKmV"
            >
              Fazer Login
            </button>
          </div>
        </form>
      </div>
      <div
        class="styles__ContentCol-sc-f3t2w6-3 gtvXPC"
        style="box-sizing: border-box; min-height: 1px; position: relative; padding-left: 15px; padding-right: 15px; width: 50%; flex-basis: 50%; flex-grow: 0; flex-shrink: 0; max-width: 50%; margin-left: 0%;"
      >
        <div
          class="styles__ContentContainer-sc-f3t2w6-4 jTaLgu"
        >
          <img
            alt="Correta.app"
            src="/images/home.png"
          />
          <span>
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 448 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
              />
            </svg>
            <h1
              class="sc-gKAaRy eeYKHx"
            >
              facilidade
            </h1>
          </span>
          <span>
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 448 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
              />
            </svg>
            <h1
              class="sc-gKAaRy eeYKHx"
            >
              resultados
            </h1>
          </span>
          <p
            class="sc-iemWCZ bCVkDu"
          >
            Tenha acesso e controle de seus cliente, melhorando seu processo de vendas e gerando mais resultados.
          </p>
        </div>
      </div>
    </div>
    <span
      style="display: table; clear: both;"
    />
  </div>
</div>
`;
